Allow the GRPC Protobuffers example to compile, add toJson support

Downchuck opened the pull request:
This adds .toJson() for easy serialization to string and support for generating interfaces for basic service definitions.

Looking for code review and/or acceptance on this patch before working on additional flags for code generation on method calls and config flags for skipping "null" values in JSON serialization.

msoucy commented on import/dproto/intermediate.d:
```diff
@@ -267,3 +291,69 @@ struct Field {
+			case 'p':
+				sink.formattedWrite("service %s {\n", name);
+				break;
+			case 'i':
```
`"%i"` meaning what? The names that I had in place are `"%d"` -> D, `"%p"` -> proto

Downchuck replied:
%i - "implemented class". At present, this creates a class with stub
methods, for quick testing/rapid prototyping.
It's not as graceful as I'd like... I'd rather pass flags, but it's hacky
to use precision/width on settings like this.

Do you have some suggestions on how we might send flags?
FormatSpec seems to hold a single char, as well as width/precision
arguments.

It seemed like we could use those to specify additional options to the
toString methods.
For instance: %3p could mean "output protobuf version 3".

Alternatively, if we keep a context variable around, all pointing to a
parent class, then the toString methods
could simply access that for flag info (though this makes them more tightly
coupled).

I'm going to have a similar issue with JSON. It's just a minor tweak (much
like implemented class), but the flags for it would be weird.

-Charles



On Thu, Mar 19, 2015 at 7:48 AM, Matt Soucy <notifications@github.com>
wrote:

> In import/dproto/intermediate.d
> <https://github.com/msoucy/dproto/pull/40#discussion_r26760384>:
>
> > +					sink.formattedWrite("%s %s (%s); { %s res; return res; }\n", responseType, name, requestType);
> > +					break;
> > +				default:
> > +					sink.formattedWrite("%s %s (%s);\n", responseType, name, requestType);
> > +					break;
> > +			}
> > +		}
> > +	}
> > +
> > +	const void toString(scope void delegate(const(char)[]) sink, FormatSpec!char fmt)
> > +	{
> > +		switch(fmt.spec) {
> > +			case 'p':
> > +				sink.formattedWrite("service %s {\n", name);
> > +				break;
> > +			case 'i':
>
> "%i" meaning what? The names that I had in place are "%d" -> D, "%p" ->
> proto
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/msoucy/dproto/pull/40/files#r26760384>.
>

msoucy replied:
I'm fond of the idea of using the FormatSpec to carry around information like this - it is, after all, really just changing the behavior of the `toString` functions.

Naturally, if we do this we'd have to document the possible flags for each print format, and possibly refactor a bit, but I think it has potential. Great example with `"%3p"`!

Downchuck replied:
Using %d makes the most sense here, could use width/precision to pass additional flags -- %d being service with interface, %1d being service with stub implementation, %2d being service with RPC stub implementation (they call controller.SetFailed and are the default in C++ code gen I believe).

I'm slightly worried about scope creep of service generation and what the default behavior should be.

Downchuck replied:
It seems like RPC stub implementation should be the default, given that it's the default for C++. The other options are handy -- I think that ".1" (implementation only) might even be a useful flag for message, only returning a dumb primitive struct, instead of an implementation.

msoucy commented on import/dproto/intermediate.d:
```diff
@@ -267,3 +291,69 @@ struct Field {
+					}
+					break;
+				case 'i':
+					sink.formattedWrite("%s %s (%s); { %s res; return res; }\n", responseType, name, requestType);
```
Semicolon printed on this line?

Downchuck replied:
Oops. Will fix in next patchset.

msoucy commented:
SYN - status?

msoucy commented on import/dproto/intermediate.d:
```diff
@@ -216,6 +236,10 @@ struct Field {
 		switch(fmt.spec) {
 			case 'p':
+				if(fmt.width == 3)
+				sink.formattedWrite("%s %s = %s%p; ",
```
Where did you find this info about the new format?

Downchuck replied:
I don't understand the question: when you say "format", what do you mean?

msoucy replied:
The differences in proto2 and proto3 syntax

Downchuck replied:
optional/required is no longer in proto3 -- it's in the change log:
https://github.com/google/protobuf/releases

That said, repeated is still in there, so it sounds like I have a bug.

msoucy replied:
Indeed. I looked at some of their code in the release, some of it uses `optional` even with `syntax = "proto3"` and some doesn't...

Downchuck replied:
I think it's the "required" tag they took out: "removal of required fields"

Downchuck commented:
Added precision for code generation flags on services, for interface, void style interface and stub method (1, 2, 3). Added "version 1" as a width, as version 1 does not exist in protobufs. It just generates plain D code, without the serialization/deserialization and protobuffer objects.

msoucy commented:
I won't be able to really look at this today, but a quick overview made me want to discuss some things about the design. I'll post a bit more about it tomorrow, but for now I'll say that I'm thinking about [flPlus](http://dlang.org/phobos/std_format.html#.FormatSpec.flPlus) (or flMinus) for denoting whether stubs should be generated, and I'd like to redesign the printing to use the Visitor design pattern - make things a little more convenient for growth in the future.

Downchuck commented:
got a little stung by this one: (format moved from string to std.format recently)
https://github.com/D-Programming-Language/phobos/pull/2732

msoucy commented:
Same issue from [this little problem](https://github.com/msoucy/dproto/issues/43#issuecomment-87504226)

Downchuck commented:
I'd like to use flMinus to strip stubs (instead of using width "1"), and address that std.format issue by just importing all of std.format and std.string where needed. Will that wrap up this PR? I'd like to move onto implementing "oneof". I agree that a visitor pattern would make for better flexibility and allow for the code generation to be more extensible.

-Charles

msoucy commented:
Yeah, that's what I had in mind. Thanks!

Downchuck commented:
Are there any remaining issues blocking this patch request?

msoucy commented:
Sorry, I've been extremely ill recently.

Stylistic: please put curly braces around if and else blocks, even if they're only one line

I don't know enough about proto3 to know if it's proper, but having the D generator specify version 3 seems a bit odd to me,  does it need that?

Other than those, if you can link to done test code to try it with, it looks pretty good to me

denizzzka commented:
Interested in this too.
Wish: please add example to "example" and into README.

Downchuck commented:
Required fields were taken out of Proto3 - but that said, we can still run the logic.

I've added an example to README. I'd be happy to add some examples to example in a subsequent revision supporting some semblance of rpc "stream".

msoucy commented:
Yes, they were removed from proto3, but dproto should still support them at this point in time - particularly since proto3 isn't really released yet. (They recently added a [proto3 language guide](https://developers.google.com/protocol-buffers/docs/proto3), so it appears that there's an official release planned soon?)  We can't just break what compatibility we have with 2 quite yet, as enticing as 3 is.

Please add the example as a unit test, probably using `__traits(compiles, ...)`

denizzzka commented:
> We can't just break what compatibility we have with 2 quite yet, as enticing as 3 is.

Looks like template argument should be implemented in near future for switching between two versions of protocol

msoucy commented:
The wire protocol seems to be identical between the two, the main differences are in things like required fields being removed, mapped types added (but that's backward-compatible with the existing system, and proto2 can receive the same data as a `repeated {key=1; value=2}` type of structure), the `Any` type (which I still have to look at for implementation), etc. So the wire protocol itself should be identical and shouldn't need a template switch. The only real changes should be to the parser.

denizzzka commented:
Oh, yep

denizzzka commented:
What about fromJson?

Downchuck commented:
Fixed the example, something is a little weird with picking up on single value types; I'd think Typename("string") would work fine in place of Typename var; var.onlyarg ="string";

msoucy commented:
Sorry, I'm not quite sure what you mean. Could you elaborate?

Downchuck commented:
I received a template error when trying to use the struct literal syntax, possibly because I was using it incorrectly:
http://dlang.org/struct.html#struct-literal

msoucy commented:
So, I have some work on the `uda` branch that would (ideally) render that not an issue, but it might be a ways away.

Downchuck commented:
Anything more needed for this pull request?

msoucy commented:
Change the "Protocol Buffer 3" section to just be a headline for "Services" and I'll merge!

Reason being these changes aren't QUITE connected to proto3 specifically...

msoucy commented:
Oh dear. Think you could add some unit tests, to test if things compile and work as expected?

Also, you can do it in another commit instead of amending to the current commit...in fact I'd actually prefer it.

Downchuck commented:
Looks like that last re-base didn't go how I wanted; will try again. The patch fell pretty far behind current.

Downchuck commented:
doveralls does not handle compile-time code coverage. I added a line in there anyway, to bump coverage up, in a new commit as requested.

msoucy commented on import/dproto/buffers.d:
```diff
@@ -378,6 +400,24 @@ struct RepeatedBuffer(ulong id, string TypeString, RealType, bool isDeprecated=f
 		}
 	}
 
+	/** serialize as JSON @todo */
```
If this is done, please remove the `@todo`
